cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -212,20 +218,37 @@ func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathM
 		return true, fmt.Errorf("fail to get etcd pod's hash: %v", err)
 	}
 
-	// Write the updated etcd static Pod manifest into the temporary directory
+	// Write the updated etcd static Pod manifest into the temporary directory, at this point no etcd change
+	// has occured in any aspects.
 	if err := etcdphase.CreateLocalEtcdStaticPodManifestFile(pathMgr.TempManifestDir(), cfg); err != nil {
-		return true, rollbackEtcdData(cfg, fmt.Errorf("error creating local etcd static pod manifest file: %v", err), pathMgr)
+		return true, fmt.Errorf("error creating local etcd static pod manifest file: %v", err)
 	}
 
 	// Perform etcd upgrade using common to all control plane components function
 	if err := upgradeComponent(constants.Etcd, waiter, pathMgr, cfg, beforeEtcdPodHash, recoverManifests); err != nil {
-		return true, rollbackEtcdData(cfg, err, pathMgr)
+		// Since etcd upgrade component failed, the old manifest has been restored
+		// now we need to check the heatlth of etcd cluster if it came back up with old manifest
+		if _, err := etcdCluster.GetEtcdClusterStatus(); err != nil {
+			return true, fmt.Errorf("fatal error upgrading local etcd cluster: %v, the backup of etcd database is stored here:(%s)", err, backupEtcdDir)
+		}
+		// Since etcd cluster came back up with the old manifest, it is not a fatal failure in theory the uprgade can proceed further
+		return false, fmt.Errorf("non-fatal error upgrading local etcd cluster: %v", err)
 	}
 
 	// Checking health state of etcd after the upgrade
-	etcdStatus, err = etcdCluster.GetEtcdClusterStatus()
-	if err != nil {
-		return true, rollbackEtcdData(cfg, fmt.Errorf("etcd cluster is not healthy after upgrade: %v rolling back", err), pathMgr)
+	if _, err = etcdCluster.GetEtcdClusterStatus(); err != nil {
+		// Despite the fact that upgradeComponent was sucessfull, there is something wrong with etcd cluster
+		// First step is to rollback to the old etcd manifest
+		if err := rollbackOldManifests(recoverManifests, err, pathMgr, true); err != nil {
+			return true, fmt.Errorf("fatal error upgrading local etcd cluster: %v, the backup of etcd database is stored here:(%s)", err, backupEtcdDir)
+		}
+		// Since rollback of the old etcd manifest was successful, checking again the status of etcd cluster
+		if _, err := etcdCluster.GetEtcdClusterStatus(); err != nil {
+			// Nothing else left to try to recover etcd cluster
+			return true, fmt.Errorf("fatal error upgrading local etcd cluster: %v, the backup of etcd database is stored here:(%s)", err, backupEtcdDir)
+		}
+		// Since etcd cluster came back up with the old manifest, it is not a fatal failure in theory the uprgade can proceed further
+		return false, fmt.Errorf("non-fatal error upgrading local etcd cluster: %v", err)
```

Comment:
let's consider this a **fatal error** (quit the general upgrade as the etcd version would be wrong if we can't upgrade it)
The purpose of the rollback here and re-check is to avoid disruption for the user (the upgrade failed but the state was preserved)

Reply:
`s/non-fatal error upgrading local etcd cluster/fatal error when trying to upgrade the etcd cluster: %v. Rolled the state back to pre-upgrade state./`

Reply:
done, but then we will not have any non-fatal cases unless the complete upgrade is successful, now the question is why do we need fatal/non-fatal return?

Reply:
the first case is non-fatal